From 72789601b55d103a28b6e8e259f3db1d31af0289 Mon Sep 17 00:00:00 2001
From: Vít Kazík <catelixor@proton.me>
Date: Thu, 30 Jul 2026 02:26:50 +0200
Subject: [PATCH] kernel: mtd: spinand: add support for Micron MT29F1G01AAADD
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Add support for the Micron MT29F1G01AAADD (M68A) SPI NAND chip.

Signed-off-by: Vít Kazík <catelixor@proton.me>
---
 drivers/mtd/nand/spi/micron.c | 27 +++++++++++++++++++++++++++
 1 file changed, 27 insertions(+)

--- a/drivers/mtd/nand/spi/micron.c
+++ b/drivers/mtd/nand/spi/micron.c
@@ -134,6 +134,23 @@ static const struct mtd_ooblayout_ops mi
 	.free = micron_4_ooblayout_free,
 };
 
+static int micron_m68a_ooblayout_free(struct mtd_info *mtd, int section,
+				      struct mtd_oob_region *region)
+{
+	if (section >= 4)
+		return -ERANGE;
+
+	region->offset = (section * 16) + 2;
+	region->length = 6;
+
+	return 0;
+}
+
+static const struct mtd_ooblayout_ops micron_m68a_ooblayout = {
+	.ecc = micron_4_ooblayout_ecc,
+	.free = micron_m68a_ooblayout_free,
+};
+
 static int micron_select_target(struct spinand_device *spinand,
 				unsigned int target)
 {
@@ -404,6 +421,16 @@ static const struct spinand_info micron_
 		     SPINAND_ECCINFO(&micron_8_ooblayout,
 				     micron_8_ecc_get_status),
 		     SPINAND_SELECT_TARGET(micron_select_target)),
+	/* M68A 1Gb 3.3V */
+	SPINAND_INFO("MT29F1G01AAADD",
+		     SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0x12),
+		     NAND_MEMORG(1, 2048, 64, 64, 1024, 20, 2, 1, 1),
+		     NAND_ECCREQ(4, 512),
+		     SPINAND_INFO_OP_VARIANTS(&x4_read_cache_variants,
+					      &x1_write_cache_variants,
+					      &x1_update_cache_variants),
+		     0,
+		     SPINAND_ECCINFO(&micron_m68a_ooblayout, NULL)),
 	/* M69A 2Gb 3.3V */
 	SPINAND_INFO("MT29F2G01AAAED",
 		     SPINAND_ID(SPINAND_READID_METHOD_OPCODE_DUMMY, 0x9F),
